CHANGELOG

```diff
@@ -1,6 +1,11 @@
 devel
 -----
 
+* APM-592: Added a retriable flag for queries in which, case the batch response
+  object isn't successfully received, by sending a POST request with 
+  `_api/cursor/<cursorId>/<batchId>`, the response object from the `batchd` that
+  had not been received before would be returned. The request is idempotent.
```

Comment:
Here is a suggestion:

```suggestion
* APM-592: Added a `retriable` flag for AQL queries. If enabled, a `nextBatchId` attribute is
  included in the response object of every cursor batch. Clients can use it to retrieve the
  subsequent result batch without advancing the cursor by using the new
  `POST /_api/cursor/<cursorId>/<batchId>` endpoint. This enables clients to retry receiving
  query results should a transfer fail. The request is idempotent.
```

I'm a bit confused, however. Is the client expected to remember the `nextBatchId` after receiving the first batch, then fetch the second batch normally? If It fails to retrieve the second batch, is uses the remembered ID from the first batch and retries to transfer the second batch? What if the initial batch fails already? And are cursors not destroyed as soon as the client asks for the last batch if the `retriable` option is enabled because the client might ask for that batch again? Does the server keep it around until the cursor TTL is reached?

Reply:
The first batch starts with 1, if you don't retrieve it successfully, then you won't get the `nextBatchId` returned to you. 
As the value of `nextBatchId` is deterministic and based on single increment per batch, then the user can either use the `nextBatchId` value when it's returned in the batch's response object or memorize it. 
The first batch's id value is 1. I'll add it to the documentation. 
Cursors are advanced every time you run `_api/cursor/<cursorId>` successfully and destroyed after the last batch, so you can get the in-between batches starting from the first (only the latest batch, the previous batches are not cached). 
The server keeps the latest batch cached until the next batch's response object is created, then it stores the next batch in place of the former one. The request to `_api/cursor/<cursorId>/<batchId>` doesn't advance the cursor

Reply:
Shouldn't the last batch remain cached until the cursor TTL is exceeded or the cursor is deleted manually? In the case of streaming cursors, the last batch contains additional important information. If retrieving the last batch fails, you cannot retry and need to re-run the entire query...